File: C++/7-Reverse-Integer.cpp
/*
Link of the question: https://leetcode.com/problems/reverse-integer/

7. Reverse Integer
Given a signed 32-bit integer x, return x with its digits reversed. If
reversing x causes the value to go outside the signed 32-bit integer range
[-2^31, 2^31 - 1], then return 0.
*/

//Solution

class Solution {
public:
    int reverse(int x) {


    if(x <= INT_MIN || (x >= INT_MAX - 1))
        return 0;

    int f=0; //for marking negative number

    long int ans=0;

    if(x<0)
    f=1,x=-1*x;


    for(;x!=0;x/=10)
    {
        ans=ans*10+x%10;
    }

    if(f)ans=-ans;

    if((ans<= INT_MIN) || (ans>= INT_MAX - 1))
        return 0;

    return ans;
    }
};

/*

Input: x = 123
Output: 321

Input: x = -123
Output: -321

Time Complexity: O(N) where N is the number of the digits in given number.
Space complexity: O(1)

*/

File: Python/7-Reverse_Integer.py
"""Link for the problem : https://leetcode.com/problems/reverse-integer
    
    Problem Statement :
      Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1], then return 0.
    """
 """SOLUTION :"""
    class Solution:
    def reverse(self, x: int) -> int:
        res = 0 # initiating result variable
        if(x < 0):
            a = abs(x) # application for negative inputs
            b = 0
            while a>0:
                b*=10
                b += a%10
                a//=10
            res = -b
        else:
            a = (x)
            b = 0
            while a>0:
                b*=10
                b+=a%10
                a//=10
            res = b
        if res >= 2**31 or res < -2**31 : # with respect to the ranging condition of 32 bit result
            res =0        
        return res
      
    """  Sample Input : 123
             Output : 321
            
             Input : 120
             Output : 21
              
             Input : -123
             Output : -321
              
              
       Time Complexity : O(log base 10 (n))
       Space Complexity : O(1)
        
     """
